examples/softirq-latency.bpf.c

```diff
@@ -0,0 +1,117 @@
+#include <vmlinux.h>
+#include <bpf/bpf_tracing.h>
+#include "bits.bpf.h"
+#include "maps.bpf.h"
+
+// Loosely based on https://github.com/xdp-project/xdp-project/blob/master/areas/latency/softirq_net_latency.bt
+
+// 20 buckets for latency, max range is 0.5s .. 1.0s
+#define MAX_LATENCY_SLOT 21
+
+struct softirq_latency_key_t {
+    u32 kind;
+    u64 slot;
+};
+
+struct {
+    __uint(type, BPF_MAP_TYPE_PERCPU_HASH);
+    __uint(max_entries, (MAX_LATENCY_SLOT + 1) * NR_SOFTIRQS);
+    __type(key, u32);
+    __type(value, u64);
+} softirqs_raise SEC(".maps");
+
+struct {
+    __uint(type, BPF_MAP_TYPE_PERCPU_HASH);
+    __uint(max_entries, (MAX_LATENCY_SLOT + 1) * NR_SOFTIRQS);
+    __type(key, u32);
+    __type(value, u64);
+} softirqs_entry SEC(".maps");
+
+struct {
+    __uint(type, BPF_MAP_TYPE_PERCPU_HASH);
+    __uint(max_entries, (MAX_LATENCY_SLOT + 1) * NR_SOFTIRQS);
+    __type(key, struct softirq_latency_key_t);
+    __type(value, u64);
+} softirq_entry_latency_seconds SEC(".maps");
+
+struct {
+    __uint(type, BPF_MAP_TYPE_PERCPU_HASH);
+    __uint(max_entries, (MAX_LATENCY_SLOT + 1) * NR_SOFTIRQS);
+    __type(key, struct softirq_latency_key_t);
+    __type(value, u64);
+} softirq_service_latency_seconds SEC(".maps");
+
+SEC("tp_btf/softirq_raise")
+int BPF_PROG(softirq_raise, unsigned int vec_nr)
+{
+    u64 ts = bpf_ktime_get_ns();
+    bpf_map_update_elem(&softirqs_raise, &vec_nr, &ts, BPF_ANY);
+    return 0;
+}
+
+SEC("tp_btf/softirq_entry")
+int BPF_PROG(softirq_entry, unsigned int vec_nr)
+{
+    u64 delta_us, latency_slot, *tsp, ts = bpf_ktime_get_ns();
+    struct softirq_latency_key_t latency_key = {};
+
+    tsp = bpf_map_lookup_elem(&softirqs_raise, &vec_nr);
+    if (!tsp) {
+        return 0;
+    }
+
+    delta_us = (ts - *tsp) / 1000;
+
+    // Latency histogram key
+    latency_slot = log2l(delta_us);
+
+    // Cap latency bucket at max value
+    if (latency_slot > MAX_LATENCY_SLOT) {
+        latency_slot = MAX_LATENCY_SLOT;
+    }
+
+    latency_key.kind = vec_nr;
+    latency_key.slot = latency_slot;
+    increment_map(&softirq_entry_latency_seconds, &latency_key, 1);
+
+    latency_key.slot = MAX_LATENCY_SLOT + 1;
+    increment_map(&softirq_entry_latency_seconds, &latency_key, delta_us);
+
+    // There is some time from function start to here, but we ignore it
+    bpf_map_update_elem(&softirqs_entry, &vec_nr, &ts, BPF_ANY);
+
+    return 0;
+}
+
+SEC("tp_btf/softirq_exit")
+int BPF_PROG(softirq_exit, unsigned int vec_nr)
+{
+    u64 delta_us, latency_slot, *tsp, ts = bpf_ktime_get_ns();
```

Comment:
Most of the function body here and above can be abstracted away into a function.